### ISSUE TITLE
Question: Multipart form-data with Refit

### ISSUE THREAD

**Issue by tranceboyjohny** [reactiveui/refit#930]
Hi, 

I have an dotnet core web api endpoint that takes a strongly typed object as argument
```
public class FileUploadArgs
{   
    public int Id { get; set; }
    public string Email { get; set; }
    public IFormFileCollection Files { get; set; }
}
```

Endpoint signature:
```
[HttpPost, Route("upload")]
public async Task<IActionResult> UploadFile([FromForm] FileUploadArgs args)
```

I call this method via HTTPIE on shell as
```
$ http -f POST localhost:5000/api/v3/file/upload Id='123456' Email='test@test.com' 'Files'@/mnt/c/apiloadtest/2124545.pdf x-api-key:'{my-api-key}'
```

I want to write this endpoint in Refit interface in another app that will send files to this endpoint along with other form data.

I'm not sure how this can be be written in Refit. Any help would be appreciated!

Thanks!

**Comment by ThaisAbreuCarvalho** [comment]
I had just one file to send but I sent it this way:

 [Multipart]
 [Post("your route")]
 Task Populate([AliasAs("parametername")] byte[] your file);

**Comment by xiaolei000129** [comment]
Hi 
I also encountered this problem, how do you solve this problem

**Comment by xiaolei000129** [comment]
Help the baby. The baby's crying

**Comment by azhdarshirinzada** [comment]
> 

send files as byte array or base64 string

**Comment by azhdarshirinzada** [comment]
> Hi I also encountered this problem, how do you solve this problem

send files as byte array or base64 string

**Comment by xiaolei000129** [comment]
> But it lost its original meaning And then there's no way to get the file information across

**Comment by azhdarshirinzada** [comment]
you have a second option to wait when refit devs add this feature. They didn't it for a 2 years FYI))

**Comment by xiaolei000129** [comment]
> 
Mm-hmm. Ooooooo the baby is already crying

**Comment by xiaolei000129** [comment]
![image](https://user-images.githubusercontent.com/44334120/199142132-c65198cc-05fd-43fe-8507-3f421fea3ec0.png)

I successfully fulfilled the requirement

**Comment by xiaolei000129** [comment]
![image](https://user-images.githubusercontent.com/44334120/199142589-41f84d77-dcf7-441a-a08f-9cf781bd918d.png)

I successfully fulfilled the requirement

**Comment by TRAD-Anthony-CKO** [comment]
While multipart works, sending `form-data` in the request is not supported. The content-type is always a JSON internally. Any workaround this?

**Comment by Piedone** [comment]
I'm also struggling with this under https://github.com/OrchardCMS/OrchardCore/issues/16430. I couldn't find a way to send a request with Refit to the below API endpoint:

```c#
[HttpPost]
[Route("setup")]
public async Task<ActionResult> Setup(SetupApiViewModel model, IFormFile recipe = null)
```

**Comment by glennawatson** [comment]
Multipart form-data is supported — decorate the method with `[Multipart]` and pass `StreamPart`/`ByteArrayPart`/`FileInfo`/`Stream`/`byte[]` parameters (see README "Multipart uploads"). Sending a file inside a larger JSON object is an ASP.NET model-binding concern rather than a Refit feature. For usage help please use [GitHub Discussions](https://github.com/reactiveui/refit/discussions). Closing.

**Comment by Piedone** [comment]
Do you mean that you can **only** use `StreamPart`/`ByteArrayPart`/`FileInfo`/`Stream`/`byte[]` parameters? This was two years ago so I can't recall more details, but apparently [I did try](https://github.com/OrchardCMS/OrchardCore/issues/16430#issuecomment-2226019995) things like:

```c#
    [Post("/api/tenants/setup")]
    [Headers(AuthorizationBearer, RequestedWithXmlHttpRequest)]
    [Multipart]
    Task<ApiResponse<string>> SetupAsync(
        TenantSetupApiModel setupTenantParameters,
        [AliasAs("recipe")] StreamPart recipe = null);
```

**Comment by glennawatson** [comment]
Re-opening — on a closer look this is a genuine gap rather than just usage, so thanks for following up @Piedone.

In a `[Multipart]` method, a complex object parameter currently falls through to the content serializer and is added as a **single JSON part** named after the parameter (`AddMultipartItem` -> `settings.ContentSerializer.ToHttpContent(itemValue)`, `src/Refit/RequestBuilderImplementation.cs:594-602`). It is **not** flattened into individual `multipart/form-data` fields, so ASP.NET `[FromForm]` model-binding against a model like `SetupApiViewModel model` won't bind it — which is exactly what you hit.

Workaround today: pass each scalar as its own part (e.g. `[AliasAs("Name")] string name, ..., [AliasAs("recipe")] StreamPart recipe`) instead of one model object.

Tracking this as an enhancement: flatten a complex object's properties into individual form-data parts in multipart requests (ideally honoring `[AliasAs]`/serializer naming per property).

**Comment by glennawatson** [comment]
Sorry for the AI generated reply btw, basically just trying to triage the issues down a bit at the moment.

**Comment by Piedone** [comment]
Yeah, I figured, but appreciate the clean-up effort, thank you! (And I'm empathetic, having cleaned up about 1000 old issues in the Orchard Core issue tracker once, before AI :).)

**Comment by glennawatson** [comment]
This is addressed in the Refit 14 beta, now available on NuGet: https://www.nuget.org/packages/Refit/14.0.0-beta.1

Please give it a try - feedback ahead of the stable 14.0.0 release is very welcome.

<blockquote><img src="https://api.nuget.org/v3-flatcontainer/refit/14.0.0-beta.1/icon" width="48" align="right"><div><strong><a href="https://nuget.org/packages/Refit/">Refit 14.0.0-beta.1</a></strong></div><div>Type-safe REST API clients for .NET, generated from annotated interfaces with build-time source generation, direct request construction, System.Text.Json defaults, and support for trimming and AOT.</div></blockquote>

**Comment by Piedone** [comment]
Thank you! While I've solved my particular issue in another way already, thank you for the fix!
